apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,409 @@
+                {
+                    let busy_v = span
+                        .end_time
+                        .duration_since(span.start_time)
+                        .unwrap()
+                        .as_micros() as i64;
+                    /*
+                    let busy = span
+                        .attributes
+                        .get(&opentelemetry::Key::from_static_str("busy_ns"))
+                        .unwrap();
+                    let busy_v = match busy {
+                        Value::I64(v) => v / 1_000_000,
+                        _ => panic!("value should be a signed integer"),
+                    };
+                    */
+                    tracing::debug!("query: {}", q);
+                    tracing::debug!("busy: {}", busy_v);
+
+                    let not_found = Value::String(Cow::from("not found"));
+                    let stats = ContextualizedStats {
+                        context: Some(StatsContext {
+                            client_name: span
+                                .attributes
+                                .get(&opentelemetry::Key::from_static_str("client_name"))
+                                .unwrap_or(&not_found)
+                                .to_string(),
+                            client_version: span
+                                .attributes
+                                .get(&opentelemetry::Key::from_static_str("client_version"))
+                                .unwrap_or(&not_found)
+                                .to_string(),
+                        }),
+                        query_latency_stats: Some(QueryLatencyStats {
+                            latency_count: vec![busy_v],
+                            request_count: 1,
+                            ..Default::default()
+                        }),
+                        ..Default::default()
+                    };
+                    let operation_name = span
+                        .attributes
+                        .get(&opentelemetry::Key::from_static_str("operation_name"));
+                    // XXX NEED TO NORMALISE THE QUERY
+                    let key = normalize(operation_name, &q.as_str());
+
+                    let msg = reporter
+                        .submit_stats(graph.clone(), key, stats)
+                        .await
+                        .map_err::<TraceError, _>(|e| e.to_string().into())?
+                        .into_inner()
+                        .message;
+                    tracing::info!("server response: {}", msg);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn normalize(op: Option<&opentelemetry::Value>, q: &str) -> String {
+    // If we don't have an operation name, no point normalizing
+    // it. Just return the unprocessed input.
+    let op_name: String = match op {
+        Some(v) => v.as_str().into_owned(),
+        None => return q.to_string(),
```

Comment:
not sure what's going on here

Reply:
It's possible to have an operation which isn't named. In this case, just return the supplied data as a string.

Reply:
Right; as described below, the expected behavior for this case is to do everything the same and put `-` as the first-line operation name. (Note that if a GraphQL document has an operation without a name, it must be the only operation (otherwise this is a validation error) and you may not specify a top-level `operationName`.)

Reply:
I've cleaned this code up to use "-" if there is no operation name and then attempt to follow the rules you have outlined. I'm not 100% sure that I've done the right thing, so I won't mark this as resolved for now.